torch/csrc/autograd/engine.cpp

```diff
@@ -623,39 +653,56 @@ auto Engine::execute(const edge_list& roots,
   // Lock post_callbacks_lock_ before clearing final_callbacks_
   ClearCallbacks _cb_guard(final_callbacks_, post_callbacks_lock_);
 
-  GraphTask graph_task(keep_graph, create_graph);
+  std::shared_ptr<GraphTask> graph_task = std::make_shared<GraphTask>(keep_graph, create_graph);
   // Lock mutex while GraphTask is being set up
-  std::unique_lock<std::mutex> lock(graph_task.mutex_);
+  std::unique_lock<std::mutex> lock(graph_task->mutex_);
 
   // Now compute the dependencies for all executable functions and queue the root
   auto graph_root = std::make_shared<GraphRoot>(roots, inputs);
-  compute_dependencies(graph_root.get(), graph_task);
+  compute_dependencies(graph_root.get(), *graph_task);
   if (!outputs.empty()) {
-    graph_task.init_to_execute(*graph_root, outputs);
+    graph_task->init_to_execute(*graph_root, outputs);
   }
-  ready_queue(at::kCPU).push(FunctionTask(&graph_task, std::move(graph_root), InputBuffer(0)));
+  ready_queue(at::kCPU).push(FunctionTask(graph_task, std::move(graph_root), InputBuffer(0)));
 
   // Not a worker
   if (worker_device == NO_DEVICE) {
     // Wait for all tasks to complete
-    graph_task.not_done_.wait(lock, [&graph_task]{
-      return graph_task.outstanding_tasks_.load() == 0;
+    graph_task->not_done_.wait(lock, [&graph_task]{
+      return graph_task->outstanding_tasks_.load() == 0;
     });
   } else {
     // Get back to work while we wait for our new graph_task to
     // complete!
     // See Note [Reentrant backwards]
-    graph_task.owner_ = worker_device;
-    lock.unlock();
-    thread_main(&graph_task);
+    // graph_task.owner_ = worker_device;
+    // lock.unlock();
+    // thread_main(&graph_task);
+    // If no extra threads remaining, create a new thread for reentrant call
+    if (!thread_pool_shared_->num_workers_.load()) {
```

Comment:
Hmm... isn't there a race here?

1. Threads 1 and 2 want to do reentrant backwards. The thread pool is empty
2. Thread 1 sees the thread pool is empty, allocates a new thread
3. Thread 2 sees there is a thread, and pushes its graph task to work on it, consuming the thread that thread 1 allocated
4. Thread 1 tries to push a graph task to its thread, but there are no more threads! Deadlock.

Reply:
I thought of this race because I was skeptical about whether or not `num_workers_` should be atomic.